subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1602 @@
+			if (wlk_ate.len == 0U) {
+				/* skip delete entry as it is already the
+				 * last one
+				 */
+				return 0;
+			}
+		} else if (len == wlk_ate.len) {
+			/* do not try to compare if lengths are not equal */
+			/* compare the data and if equal return 0 */
+			if (len <= 4) {
+				rc = memcmp(&wlk_ate.data, data, len);
+				if (!rc) {
+					return 0;
+				}
+			} else {
+				rc = zms_flash_block_cmp(fs, rd_addr, data, len);
+				if (rc <= 0) {
+					return rc;
+				}
+			}
+		}
+	} else {
+		/* skip delete entry for non-existing entry */
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	/* calculate required space if the entry contains data */
+	if (data_size) {
+		/* Leave space for delete ate */
+		required_space = data_size + ate_size;
+	}
+
+	k_mutex_lock(&fs->zms_lock, K_FOREVER);
+
+	gc_count = 0;
+	while (1) {
+		if (gc_count == fs->sector_count) {
+			/* gc'ed all sectors, no extra space will be created
+			 * by extra gc.
+			 */
+			rc = -ENOSPC;
+			goto end;
+		}
+
+		if (fs->ate_wra >= (fs->data_wra + required_space)) {
+
+			rc = zms_flash_write_entry(fs, id, data, len);
+			if (rc) {
+				goto end;
+			}
+			break;
+		}
+		rc = zms_sector_close(fs);
+		if (rc) {
+			goto end;
+		}
+
+		rc = zms_gc(fs);
+		if (rc) {
+			goto end;
+		}
+		gc_count++;
+	}
+	rc = len;
+end:
+	k_mutex_unlock(&fs->zms_lock);
+	return rc;
+}
+
+int zms_delete(struct zms_fs *fs, uint32_t id)
+{
+	return zms_write(fs, id, NULL, 0);
+}
+
+ssize_t zms_read_hist(struct zms_fs *fs, uint32_t id, void *data, size_t len, uint32_t cnt)
+{
+	int rc, previous_sector_num = -1;
+	uint64_t wlk_addr, rd_addr;
+	uint8_t current_cycle;
+	uint32_t cnt_his;
+	struct zms_ate wlk_ate;
+	size_t ate_size;
+#ifdef CONFIG_ZMS_DATA_CRC
+	uint32_t computed_data_crc;
+#endif
+
+	if (!fs->ready) {
+		LOG_ERR("zms not initialized");
+		return -EACCES;
+	}
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	if (len > (fs->sector_size - 2 * ate_size)) {
+		return -EINVAL;
+	}
+
+	cnt_his = 0U;
+
+#ifdef CONFIG_ZMS_LOOKUP_CACHE
+	wlk_addr = fs->lookup_cache[zms_lookup_cache_pos(id)];
+
+	if (wlk_addr == ZMS_LOOKUP_CACHE_NO_ADDR) {
+		rc = -ENOENT;
+		goto err;
+	}
+#else
+	wlk_addr = fs->ate_wra;
+#endif
+	rd_addr = wlk_addr;
+
+	while (cnt_his <= cnt) {
+		rd_addr = wlk_addr;
+		rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+		if (rc) {
+			goto err;
+		}
+		if (wlk_ate.id == id) {
+			/* read the ate cycle only when we change the sector
+			 * or if it is the first read
+			 */
+			rc = zms_get_cycle_on_sector_change(fs, rd_addr, previous_sector_num,
+							    &current_cycle);
+			if (rc) {
+				return rc;
+			}
+			if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+				cnt_his++;
+			}
+			previous_sector_num = SECTOR_NUM(rd_addr);
+		}
+		if (wlk_addr == fs->ate_wra) {
+			break;
+		}
+	}
+
+	if (((wlk_addr == fs->ate_wra) && (wlk_ate.id != id)) || (wlk_ate.len == 0U) ||
+	    (cnt_his < cnt)) {
+		return -ENOENT;
+	}
+
+	if (len <= 4) {
```

Comment:
When the data size is small enough, the data itself is stored in place of the data CRC, directly in the ATE.
In such case, the ATE data offset field should also not be used ? It may then store data too, increasing the amount of data stored into the ATE from 4 to 8 ?

Reply:
Actually I thought about that. It is going then to be a little bit complex by using memcpy sometimes to set/get data and concatenate the result.
I would prefere keeping it simple then in later versions, when I introduce the multiple format ATE we could for example have a format with 8 bytes or more data within ATE

Reply:
@RICCIARDI-Adrien please have a look at the last version, it is implementing this

Reply:
This is nice !